### DISCUSSION TITLE
Swappable SQL Dialects

### DISCUSSION THREAD

**Discussion by paullatzelsperger** [eclipse-edc/Connector#923] [Ideas]
During the implementation of various SQL issues it emerged that it might pay off to make the SQL dialect exchangeable. The reason for this is, that different dialects offer sometimes a more neat and succinct syntax, and can even reduce the number of DB queries necessary.

For example in ANSI SQL there is no `upsert`, we'd have to do a `SELECT` followed by a `UPDATE` or an `INSERT`, whereas some dialects offer [specialized statements for this](https://stackoverflow.com/a/15252360/7079724).

Another example where this could come in handy is the ability to have explicit JSON type columns in Postgres.

Using the example of, say, the `SqlContractNegotiationStore`, we could have a wrapper class `SqlDialect` that gets passed in through the CTor:

```java
public class SqlContractNegotiationStore implements ContractNegotiationStore {
    //...
    public SqlContractNegotiationStore(/*other params*/, SqlDialect dialect){
        this.dialect = dialect;
    }
    // ...
    
    public void save(ContractNegotiation negotiation){
         transactionContext.execute(() -> {
            try (var connection = getConnection()) {
                for (var definition : definitions) {
                    executeQuery(connection, dialect.getUpsertTemplate(),
                            negotiation.getId(),
                            /*other args*/);
                }
            } catch (Exception e) {
                throw new EdcPersistenceException(e.getMessage(), e);
            }
        });
   
    }
}
```

For the `SqlDialect` interface there could be multiple implementations, for example:
```java
public interface SqlDialect {
   String getUpsertTemplate();
   String getDeleteTemplate();
   // ...others
}
```
```java
public class PostgresDialect implements SqlDialect{

    @Override
    public String getUpsertTemplate(){
        return "INSERT INTO ... ON CONFLICT ..."; //actual statement needs to be detailed!
    }

    @Override
    public String getDeleteTemplate(){
        return "DELETE * FROM %s WHERE %s RETURNING *"; //returns the deleted row
    }
}
```

By default I'd suggest targeting either ANSI/ISO SQL, or PostgreSQL and prepare a way of providing other implementations for `SqlDialect`.

_From a timing perspective I'd recommend doing this post Milestone 3._

**Comment by MoritzKeppler** [comment]
I'm fine with limiting the scope to PostgreSQL. There are public cloud offerings for Postgres, on premises solutions and also other db products using the same protocol (e.g. CockroachDB). The Cosmos DB implementation is also product specific.
If we realize that another SQL solution is needed, we can still introduce an abstraction layer (or write extensions that use a ORM mapper which already provides an abstraction layer).
<sub> Moritz Keppler <moritz.keppler@mercedes-benz.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/mercedes-benz/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

**Reply by paullatzelsperger** [reply]
The CosmosDB impl is a bit of a different animal, because it doesn't use JDBC but its own client SDK and it's not a relational database. However, re-using some stuff between CosmosDB and "conventional" SQL backends is very possible, eg. the fluent mechanism to generate `SELECT` statements, i.e. `Clause` API. 

My point was to _have the dialect exchangeable_ and provide Postgres as default, not to initially provide a myriad of dialects. 

Also, EDC being a platform, extensions with specific OR mappers would likely go into the downstream project's repo, no?

**Comment by jimmarino** [comment]
+1 on introducing this and focusing on Postgres as a first implementation. From an EDC perspective, we should support multiple SQL backends.

**Comment by mokhairymahmoud** [comment]
Is there any intention to revive this initiative ? Also, can another DB (ex: MariaDB) support be implemented as an extension or it requires big changes in the EDC core ?

**Reply by ndr-brt** [reply]
I don't think there's a real interest in this as the long term plan could be to have a "vanilla EDC control-plane distribution" that will be based on PostgreSQL.
Generally speaking there's no fancy SQL stuff in EDC, apart from the JSON data type management.
At worst, there will be the need to re-write all the `*Store` classes to support MariaDB.

EDC Committer group won't sponsor this in any case.
